### PR TITLE
fix: use unique_ptr for geometry in DD4hep_service, hand out not_null const pointers

### PR DESCRIPTION
**Pull request by wdconinc**
### Briefly, what does this PR introduce?
While looking into #1032, #1035, #1036 I started realizing two things:
- cell ID position converters are created by algorithms from geometry instead of doing this once in the service,
- there is no const protection in the detector pointers passed to algorithms (i.e. an algorithm could change geometry),
- there is no protection against null pointers being passed from geometry.

The first two points are related since the interface for creating a cell ID position converter requires a non-const detector (see also https://github.com/AIDASoft/DD4hep/pull/1148).

This PR centralizes the creation of cell ID position converters, clarifies ownership through unique_ptrs, and makes sure no null pointers are passed to the algorithms. In places where the constructor gets the const not_null pointer, we can also impose this in the class on the relevant member variable.

### What kind of change does this PR introduce?
- [x] Bug fix (issue: avoid creation of cell ID position converters in algorithms)
- [ ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators:
  - @simonge reordered MatrixTransferStatic init arguments
  - @c-dilks @chchatte92 this also affects RichGeo_service etc.

### Does this PR introduce breaking changes? What changes might users need to make to their code?
Yes, now requires GSL (which has always been in eic-shell and is a requirement).

### Does this PR change default behavior?
No.

### CODE REVIEW

**Comment by veprbl** [comment]
Do we really want to smear non-null check across our interfaces? Is it not enough to just ensure that the geometry service doesn't return null.

**Comment by wdconinc** [comment]
> Do we really want to smear non-null check across our interfaces? Is it not enough to just ensure that the geometry service doesn't return null.

It's heavy on notation, I admit, but the requirement for a not_null pointer is in this case imposed by the algorithms, not by the service. The DD4hep_service can satisfy the demand by promising to return a not_null pointer. Putting the requirement in the algorithms makes clear to anyone working in those algorithms that `detector` can be relied upon to be not_null (yeah, this benefit it kinda negated by the whole init-function and clearer in the RichGeo_service).

**Comment by wdconinc** [comment]
@veprbl I removed the gsl::not_null on all algorithms init calls in https://github.com/eic/EICrecon/pull/1045/commits/b9b97f4ccbba7bb7a20c9c3ae27a44701e06232b

This doesn't really add much utility on the init calls if we can't set it on the class members. I've kept it in DD4hepBfield since there we construct with the detector geometry provided as constructor argument (similar to RichGeo_services).

**Comment by c-dilks** [comment]
The dRICH benchmarks look okay with this PR.

File: src/algorithms/calorimetry/CalorimeterHitDigi.h [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
 #pragma once
 
+#include <gsl/gsl>
```
Comment (veprbl) [inline src/algorithms/calorimetry/CalorimeterHitDigi.h:16]: Unused? (and ditto for other occurrences)

File: src/services/geometry/dd4hep/DD4hep_service.cc [hunk 2]
```diff
@@ -30,21 +29,22 @@ DD4hep_service::~DD4hep_service(){
...
 /// Return pointer to the cellIDPositionConverter object.
 /// Call Initialize if needed.
 //----------------------------------------------------------------
-std::shared_ptr<const dd4hep::rec::CellIDPositionConverter>
-DD4hep_service::cellIDPositionConverter() {
+gsl::not_null<const dd4hep::rec::CellIDPositionConverter*>
+DD4hep_service::converter() {
     std::call_once(init_flag, &DD4hep_service::Initialize, this);
-    return (m_cellid_converter);
+    return (m_cellid_converter.get());
```
Comment (veprbl) [inline src/services/geometry/dd4hep/DD4hep_service.cc:47]:
```suggestion
    return m_cellid_converter.get();
```

**Comment by veprbl** [comment]
> @veprbl I removed the gsl::not_null on all algorithms init calls in [b9b97f4](https://github.com/eic/EICrecon/commit/b9b97f4ccbba7bb7a20c9c3ae27a44701e06232b)
> 
> This doesn't really add much utility on the init calls if we can't set it on the class members. I've kept it in DD4hepBfield since there we construct with the detector geometry provided as constructor argument (similar to RichGeo_services).

Yes. That was my point, that it's probably the return values where gsl::not_null makes sense (I don't claim any deep philosophical understanding of GSL, though).

File: src/algorithms/calorimetry/CalorimeterHitDigi.h (hunk 1 above)
Comment (wdconinc) [reply src/algorithms/calorimetry/CalorimeterHitDigi.h:16]: Yeah, didn't include that in my sed. Let me add (subtract) them.

File: src/services/geometry/dd4hep/DD4hep_service.cc [hunk 3]
```diff
@@ -30,21 +29,22 @@ DD4hep_service::~DD4hep_service(){
 /// Return pointer to the dd4hep::Detector object.
 /// Call Initialize if needed.
 //----------------------------------------------------------------
-dd4hep::Detector* DD4hep_service::detector() {
+gsl::not_null<const dd4hep::Detector*>
+DD4hep_service::detector() {
     std::call_once(init_flag, &DD4hep_service::Initialize, this);
-    return (m_dd4hepGeo);
+    return (m_dd4hepGeo.get());
```
Comment (wdconinc) [inline src/services/geometry/dd4hep/DD4hep_service.cc:35]:
```suggestion
    return m_dd4hepGeo.get();
```

**Review by veprbl — APPROVED** [review]
Diff LGTM, capybara happy

**Pull request merged**